[Jake] TODO-HTTP 미션 제출

cocahack opened the pull request:
TODO 백엔드 완성 후 테스트까지 마친 뒤 PR 보냅니다. 테스트 커버리지는 87% 입니다. [![Coverage Status](https://coveralls.io/repos/github/cocahack/nodejs-todo-web/badge.svg?branch=master)](https://coveralls.io/github/cocahack/nodejs-todo-web?branch=master)

회원가입, 로그인, 로그아웃 등 계정 관련 기능과 todo, todolist의 조회, 삽입, 삭제, 갱신 등 기본적인 기능을 구현되었습니다.

다만 라우터와 미들웨어가 제대로 구현되어 있지 않아서 callback과 if-else, try-catch가 중첩된 스파게티 코드가 만들어졌습니다.😿 
이 부분은 express.js 소스를 분석해보고 반드시 고칠 예정입니다.  💪

구현과 관련된 내용은 README.md에 업데이트했고, 앱은 Heroku로 배포해뒀습니다. 
[링크](https://todo-http.herokuapp.com/)

cocahack committed:
feat: package.json 파일 추가 및 Jest 설치

cocahack committed:
feat: Jest 모듈 추가

cocahack committed:
refactor: gitignore 수정

User-defined list 추가
- data 디렉토리

cocahack committed:
feat: eslint 모듈 및 npm script 추가

npm script는 jest와 관련된 코드

cocahack committed:
refactor: gitignore 업데이트

eslint 추가

cocahack committed:
feat: HTTP 서버 실행에 필요한 기초 코드 추가

app.js: HTTP 서버 로직, 기본적인 라우팅 포함
server.js: app.js 실행

cocahack committed:
feat: util.js 추가

파일 확장자 추출, MIME 타입 반환 등 각종 유틸성 함수를 여기에 작성

cocahack committed:
feat: init.js 추가

서버 시작 시 가장 먼저 실행할 파일
데이터베이스로 사용할 파일이 있는지 검사하고 없으면 생성

cocahack committed:
feat: init.js 테스트 추가

cocahack committed:
feat: template.js 추가

html 코드를 생성하는 파일
TODO 데이터에 따라 동적으로 코드를 생성함

cocahack committed:
feat: template.js 테스트 코드 추가

가상의 Todo, TodoList 데이터로 html 코드를 동적으로 생성한 뒤, 예상되는
html 코드와 비교

cocahack committed:
feat: Todo, Todolist 모델 추가

Todo와 Todolist를 표현하는 클래스 추가

cocahack committed:
feat: todo, todolist 테스트 추가

cocahack committed:
feat: db.js 추가

데이터베이스로 사용할 파일에 읽기, 쓰기를 수행하는 코드 작성

cocahack committed:
feat: db.js 테스트 추가

읽기 테스트는 한 번만 실행
create, update 테스트는 todo, todolist 각각 한 번 씩 수행

cocahack committed:
fix: util.js 수정

path 모듈을 불러오지 않은 문제 수정

cocahack committed:
feat: util.js 테스트 추가

cocahack committed:
feat: nodemon 모듈 추가

cocahack committed:
fix: app.js, template.js 수정

app.js
- 앱 실행 시 데이터를 읽어오지 못하는 경우 모든 요청에 대해 503 code
반환
template.js
- 동적으로 생성되는 코드의 위치가 잘못된 부분을 수정

cocahack committed:
fix: frontend js 코드 수정

최초 index.html 렌더링 과정에서 이벤트를 추가하는 코드가 없었던 문제를
수정

cocahack committed:
refactor: todo, todolist 모델 수정

id 및 position을 int로 형변환

cocahack committed:
refactor: db.js template.js 수정

todo 파일을 읽을 때 todolist 별로 나누지 않고 배열 그대로 반환하도록
변경
template.js에서 동적 html 코드를 만들 때 배열에 담긴 todo를
todolist 별로 구분한 뒤 코드를 만들도록 수정함

cocahack committed:
refactor: frontend js 수정

todo, todlist 추가 시 ajax 요청을 보내는 코드 추가

cocahack committed:
feat: todolist POST 라우터 추가

todolist 추가 시 실행될 라우터 추가

cocahack committed:
style: 배경 이미지 사이즈 조절

background-size cover 적용

cocahack committed:
feat: app.js db.js todo 업데이트 추가

app.js
- 라우터 추가
db.js
- 업데이트 시 파일을 전체 수정함

cocahack committed:
refactor: template 수정

todo li 요소에 data-id 속성 추가
ajax 요청 시 todo object를 보내기 위함

cocahack committed:
refactor: frontend js 업데이트 ajax 추가

drop 이벤트가 발생하면 ajax 요청하도록 작성
모든 todo를 전송        :ㅈ

cocahack committed:
refactor: template.js 테스트 코드 수정

바뀐 템플릿에 따라 html 코드의 기댓값을 수정

cocahack committed:
feat: 로그인 화면 추가

login.html, login.css, login.js
사용자 입력을 받아 /login path에 POST로 요청
validation 기능 필요

cocahack committed:
feat: 세션 및 쿠키 관련 유틸 함수 추가

generateSessionId
- 사용자 아이디로 임의의 문자열을 반환하는 함수
parseCookie
- 쿠키를 파싱하여 객체로 반환

cocahack committed:
refactor: 로그인 페이지 수정

비동기 로그인 구현
- fetch의 redirected가 true면 /todo로 이동
- statusCode가 403이면 로그인 에러 출력
- 그 외 상태 코드는 서버 오류 출력
UI
- 로그인 결과를 출력하는 toast 관련 코드 추가

cocahack committed:
feat: router 객체 추가

request의 method와 url로만 분기되는 간단한 router 구현

cocahack committed:
feat: sessionManager 객체 추가

사용자 세션을 관리하는 객체
- 세션 id를 키로 사용하는 map 객체를 사용
- 세션이 유효한지 체크하는 로직 존재

cocahack committed:
feat: db.js 기능 추가

사용자 정보를 읽어오는 로직 추가

cocahack committed:
refactor: app.js 수정

- if 분기 대신 router 사용
- 정적 페이지는 router에 포함하지 않고 if 분기로 처리
- 각 라우터에 세션 관련 로직 추가

cocahack committed:
refactor: 사용자에 따른 DB 접근 기능 추가

sessionmanager.js
- sessionId로 검색해  사용자 아이디를 반환하는 함수 추가
db.js
- todo, todolist의 CRU 함수의 parameter에 userId 추가
app.js
- DB 엑세스 함수 시그니처 변경과 관련된 코드 수정

cocahack committed:
refactor: 테스트 수정

db.js, todo.js, todolist.js는 app.js 테스트와 통합
util.js 테스트에 cookie parser, session 생성 관련 테스트 추가
app.js 테스트는 브라우저에서 서버에 접속할 때와 비슷한 절차로 진행

cocahack committed:
fix: 테스트 도중 발견된 버그 제거

cocahack committed:
feat: 로그인 화면 기능 추가

회원가입 버튼 추가
입력값 검증 로직 추가

cocahack committed:
feat: 회원가입 화면 추가

입력값 검증, 중복된 아이디 검사 가능

cocahack committed:
feat: db.js 기능 추가

사용자가 존재하는지 검사하는 함수 추가
회원 가입 시 회원의 디렉토리 생성 및 계정 정보와 todo 정보를 담을 파일
생성

cocahack committed:
feat: sessionManager 세션 제거 기능 추가

cocahack committed:
feat: 메인 화면 로그아웃 버튼 추가

cocahack committed:
feat: app.js 라우팅 추가

회원 가입, 로그아웃 라우터 추가
세션 만료 시, 세션 저장소에 저장된 세션을 제거하는 코드 삽입

cocahack committed:
style: 로그아웃 버튼 CSS 코드 적용

cocahack committed:
refactor: 메인 페이지 수정에 따른 template mock data 수정

cocahack committed:
feat: app.js 테스트 추가

회원가입 및 로그아웃 테스트 추가

cocahack committed:
fix: 테스트 코드에서 발견한 버그 수정

cocahack committed:
refactor: index.html 제거

cocahack committed:
refactor: 프로젝트 배포 시 서버에 설정된 포트가 설정되도록 변경

cocahack committed:
refactor: npm 스크립트 수정

node로 앱을 실행하도록 수정

cocahack committed:
refactor: init 함수 제거

init.js 파일을 제거하고 서버 시작 전 디렉토리를 확인하도록 수정

cocahack committed:
fix: 테스트 코드 버그 수정

app.js 테스트 후 데이터 삭제가 안되는 문제 수정

cocahack committed:
fix: Todo 추가 시 발생하는 버그 수정

서버에서 새로 추가한 todo, todolist의 id를 클라이언트에게 전달하도록 수정
index.html을 서버에서 렌더링 할 때 todolist의 id가 반영되지 않은 점 수정

cocahack committed:
feat: Todo, Todolist 삭제 버튼 front-end 코드 추가

frontend css 관련 작업만 완료
삭제 버튼 클릭 이벤트, backend에서 삭제 관련 라우터 추가, todo 및
todolist 생성 시 html 태그를 서버에서 내려주도록 변경할 예정

cocahack committed:
feat: Todo, todolist 삭제 라우터 추가

DB에 저장되는 방식은 Update와 동일
뷰에서 요청을 보낼 때 todo와 todolist를 배열로 보내야 함

cocahack committed:
feat: Todo, Todolist 삭제 버튼 이벤트 추가

버튼 클릭 시 api 호출하는 이벤트 추가
드래그&드롭, 새로운 엘리먼트 생성 시에도 이벤트를 사용할 수 있도록 기존
드롭 및 keydown 이벤트에 관련 코드 삽입

cocahack committed:
refactor: Todo, Todolist 삭제 테스트 추가 및 테스트 일부 수정

Todo, Todolist 삭제 테스트 추가
template 테스트는 app.js 테스트와 통합

cocahack committed:
fix: 로그인 버그 수정

존재하지 않는 계정으로 로그인할 때 403이 아닌 500 상태 코드가 설정되는
문제 해결

cocahack committed:
fix: app.js 테스트 코드 버그 수정

잘못된 로그인 시도와 관련된 테스트 케이스 2개 수정
객체 프로퍼티가 잘못 설정되었음

cocahack committed:
fix: 불필요한 mock data 제거

테스트 코드가 사라지면서 필요없어진 mock data 제거

cocahack committed:
refactor: README.md 수정

cocahack committed:
feat: 미들웨어 구현

express.js의 미들웨어와 유사한 기능을 가진 구현체 생성

cocahack committed:
refactor: router 리팩토링

method와 url로 함수를 호출하는 간단한 라우터에서 express.js의 미들웨어
스타일로 변경
콜백을 등록할 때, 먼저 실행할 미들웨어를 등록할 수 있으며, 이 기능은
라우터에 새로운 미들웨어 객체를 등록하여 구현함

cocahack committed:
refactor: 새로운 라우터 객체로 로직 이동

수정 사항
app.js에 몰려있던 로직들을 라우터 인스턴스 단위로 분리
request를 가공하는 로직이나 로그인 확인 관련 로직을 분리하여 미들웨어
객체에 넣거나 라우터의 콜백 앞에 오도록 변경

테스트
에러를 정상적으로 반환할 수 있는지 테스트하는 코드 추가

router.js에 문제점 발견
라우터에서 에러 발생 시 마지막 에러 처리 미들웨어로 넘어가지 않는 문제가
존재

cocahack committed:
fix: 라우터 버그 수정

라우터에서 에러 발생 시 에러 미들웨어가 실행되지 않는 문제 해결

cocahack committed:
refactor: sessionManager 객체 수정

테스트를 위해 만들었던 코드 수정

cocahack committed:
refactor: console.log 제거

cocahack committed:
feat: travis 구성 추가

honux77 commented on README.md:
```diff
@@ -1,2 +1,100 @@
+2. DB를 사용하지 않았습니다.  파일을 DB로 사용했을 때 어떤 문제점이 있을지 알기 위함입니다.
+
+### DB scheme
+
```
DB 구조 아주 좋네요! 잘했어요!

cocahack committed:
fix: Test 수정

DB 디렉토리 생성하도록 수정

cocahack committed:
feat: Coveralls 모듈 및 npm  스크립트 추가

cocahack committed:
refactor: front-end 코드 수정

fetch url 수정

cocahack committed:
fix: coveralls badge 버그 수정

honux77 commented on app.js:
```diff
@@ -0,0 +1,66 @@
+const todoRouter = require('./app/routes/todorouter');
+const todolistRouter = require('./app/routes/todolistrouter');
+
+const server = http.createServer(async (req, res) => {
```
async 키워드는 왜 있는 건가요? (자문자답) 계속 읽다보니 알겠음 ㅋㅋ

cocahack committed:
refactor: README.md 수정

미들웨어와 라우터 구현 관련 내용 삽입.
프로젝트 Todo 목록 갱신

honux77 commented on app/middleware.js:
```diff
@@ -0,0 +1,49 @@
+const Middleware = function() {
+  const middlewareContainer = [];
+  let _req, res;
```
_res인가요?

cocahack committed:
fix: README.md 오타 수정

cocahack committed:
fix: 오타 수정 및 사용하지 않는 코드 삭제

- createServer 콜백의 async 키워드 제거
- Middleware 클래스의 _res 멤버 변수에 언더바가 없었던 문제 수정
- Router 클래스에서 더 이상 사용하지 않는 클래스 및 주석 삭제

honux77 commented on app/middleware.js:
```diff
@@ -0,0 +1,49 @@
+  const _run = (index, err) => {
+    if (index < 0 || index >= middlewareContainer.length) return;
+
+    const nextMw = middlewareContainer[index];
```
nextMw 는 nextMiddleWare 군요. 가능하면 줄이지 않는 편이 좋은 것 같아요.

honux77 commented on app/middleware.js:
```diff
@@ -0,0 +1,49 @@
+    }
+  };
+
+  const _run = (index, err) => {
```
재귀 (ㅜㅜ)

cocahack replied:
시간되면 이것도 한번 바꿔보겠습니다 💪

honux77 reviewed (approved):
front는 빼고 백엔드 부분만 리뷰했어요. 깔끔하게 잘 짰습니다 👍
